crates/shared/src/sources/uniswap_v3/pool_fetching.rs

```diff
@@ -86,166 +100,365 @@ impl TryFrom<PoolData> for PoolInfo {
+            append_events(&mut checkpoint.pools, events);
+            checkpoint.block_number = new_checkpoint_block;
+        }
+        Ok(())
+    }
 }
 
 #[async_trait::async_trait]
 impl PoolFetching for UniswapV3PoolFetcher {
-    async fn fetch(&self, token_pairs: &HashSet<TokenPair>) -> Result<Vec<PoolInfo>> {
-        tracing::debug!("token_pairs {:?}", token_pairs);
-        let (mut cached_pools, outdated_pools) = self.get_cached_pools(token_pairs);
-        tracing::debug!(
-            "cached pools: {:?}, outdated pools: {:?}",
-            cached_pools,
-            outdated_pools
-        );
+    async fn fetch(
+        &self,
+        token_pairs: &HashSet<TokenPair>,
+        at_block: Block,
+    ) -> Result<Vec<PoolInfo>> {
+        let block_number = match at_block {
+            Block::Recent => self
+                .events
+                .lock()
+                .await
+                .store()
+                .last_event_block()
+                .await
+                .unwrap_or_default(),
+            Block::Number(number) => number,
+        };
 
-        if !outdated_pools.is_empty() {
-            let updated_pools = self.get_pools_and_update_cache(&outdated_pools).await?;
-            tracing::debug!("updated pools: {:?}", updated_pools);
-            cached_pools.extend(updated_pools);
+        // sanity check if events are up to date
+        let last_handled_block = self
+            .events
+            .lock()
+            .await
+            .last_handled_block()
+            .unwrap_or_default()
+            .0;
+        if block_number > last_handled_block {
+            tracing::debug!(
+                "can't get liquidity for block {} since the last handled block is {}",
+                block_number,
+                last_handled_block
+            );
+            return Ok(Default::default());
         }
 
-        Ok(cached_pools
-            .into_iter()
+        // this is the only place where this function uses checkpoint - no data racing between maintenance
+        let (mut checkpoint, checkpoint_block_number) = self.checkpoint.get(token_pairs);
+
+        if block_number > checkpoint_block_number {
+            let block_range = RangeInclusive::try_new(checkpoint_block_number + 1, block_number)?;
+            let events = self.events.lock().await.store().get_events(block_range);
+            append_events(&mut checkpoint, events);
+        }
+
+        Ok(checkpoint
+            .into_values()
             .flat_map(TryInto::try_into)
             .collect())
     }
 }
 
-#[async_trait::async_trait]
-impl Maintaining for UniswapV3PoolFetcher {
-    async fn run_maintenance(&self) -> Result<()> {
-        tracing::debug!("periodic update started");
-        let now = Instant::now();
+/// For a given checkpoint, append events to get a new checkpoint
+fn append_events(pools: &mut HashMap<H160, PoolData>, events: Vec<Event<UniswapV3Event>>) {
```

Comment:
This logic is not so simple - can we add some unit tests for it?

Reply:
Added.